File: backend/app/schemas/trip.py
"""Trip schemas (docs/03 §4 Trips). trip_code is server-generated, never client-set."""
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.models.enums import TripStatus


class TripCreate(BaseModel):
    source: str = Field(min_length=1, max_length=120)
    destination: str = Field(min_length=1, max_length=120)
    vehicle_id: uuid.UUID
    driver_id: uuid.UUID
    cargo_weight_kg: Decimal = Field(gt=0)
    planned_distance_km: Decimal = Field(gt=0)
    revenue: Decimal = Field(default=Decimal("0"), ge=0)
    notes: str | None = None


class TripComplete(BaseModel):
    end_odometer: Decimal = Field(ge=0)
    revenue: Decimal | None = Field(default=None, ge=0)
    fuel_liters: Decimal | None = Field(default=None, gt=0)
    fuel_cost: Decimal | None = Field(default=None, ge=0)

    @model_validator(mode="after")
    def _fuel_both_or_neither(self) -> "TripComplete":
        if (self.fuel_liters is None) != (self.fuel_cost is None):
            raise ValueError("Provide both fuel_liters and fuel_cost, or neither.")
        return self


class TripCancel(BaseModel):
    reason: str | None = None


class TripOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    trip_code: str
    source: str
    destination: str
    vehicle_id: uuid.UUID
    driver_id: uuid.UUID
    cargo_weight_kg: Decimal
    planned_distance_km: Decimal
    revenue: Decimal
    status: TripStatus
    start_odometer: Decimal | None
    end_odometer: Decimal | None
    notes: str | None
    created_by: uuid.UUID
    dispatched_at: datetime | None
    completed_at: datetime | None
    cancelled_at: datetime | None
    created_at: datetime
    updated_at: datetime

File: frontend/src/lib/schemas/trip.ts
import { z } from 'zod';
import { isAxiosError } from 'axios';
import type { ErrorEnvelope } from '../../types/api';

// ---------------------------------------------------------------------------
// Domain types — mirrors docs/03-API-SPEC.md (Trips/Vehicles/Drivers). These
// are not (yet) present in src/types/api.ts, so they are defined locally here
// and shared across the trips feature files.
// ---------------------------------------------------------------------------

export type TripStatus = 'draft' | 'dispatched' | 'completed' | 'cancelled';

// NOTE: money/odometer/weight NUMERIC columns are serialized by the API as
// string-safe numbers (e.g. "450.00"), not JSON numbers — confirmed live.
// Callers must `Number(...)` these before doing arithmetic/formatting.
export interface Trip {
  id: string;
  trip_code: string;
  source: string;
  destination: string;
  vehicle_id: string;
  driver_id: string;
  cargo_weight_kg: string;
  planned_distance_km: string;
  revenue: string;
  status: TripStatus;
  start_odometer: string | null;
  end_odometer: string | null;
  notes: string | null;
  dispatched_at: string | null;
  completed_at: string | null;
  cancelled_at: string | null;
  created_at: string;
}

export interface VehicleOption {
  id: string;
  registration_number: string;
  name: string;
  max_load_capacity_kg: string;
  status?: string;
}

export interface DriverOption {
  id: string;
  full_name: string;
  license_number: string;
  license_expiry: string;
  safety_score: string;
  status?: string;
}

// Pagination envelope: use `Page<T>` from '../../types/api' (owned by another
// agent, kept AS-IS) rather than duplicating it here.

export interface TripAdvisorVerdictResponse {
  verdict: 'go' | 'caution' | 'block';
  hard_failures: string[];
  risk_factors: string[];
  summary: string;
}

// ---------------------------------------------------------------------------
// zod schemas
//
// zod@4 + @hookform/resolvers@5: `z.coerce.number()` gives the schema's
// *input* type as `unknown` (coerce accepts anything), which then fails to
// satisfy `useForm<Output>()`'s single-generic assumption that TFieldValues
// equals the resolver's input type ("Type 'unknown' is not assignable to
// type 'number'"). Fix: use plain `z.number()` (input === output === number)
// and convert text -> number in each field's onChange via `valueAsNumber`.
// ---------------------------------------------------------------------------

export const tripSchema = z.object({
  source: z.string().min(2, 'Source must be at least 2 characters').max(120),
  destination: z.string().min(2, 'Destination must be at least 2 characters').max(120),
  vehicle_id: z.string().min(1, 'Vehicle is required'),
  driver_id: z.string().min(1, 'Driver is required'),
  cargo_weight_kg: z
    .number({ error: 'Cargo weight is required' })
    .positive('Cargo weight must be greater than 0'),
  planned_distance_km: z
    .number({ error: 'Planned distance is required' })
    .positive('Distance must be greater than 0'),
  revenue: z.number().min(0, 'Revenue cannot be negative').optional(),
  notes: z.string().optional(),
});

export type TripFormInputs = z.infer<typeof tripSchema>;

// end_odometer's lower bound depends on the trip's start_odometer, which is
// only known at render time — build the schema per-modal-instance.
export function buildCompleteTripSchema(startOdometer: number) {
  return z
    .object({
      end_odometer: z.number({ error: 'End odometer is required' }).positive('Must be positive'),
      revenue: z.number().min(0, 'Cannot be negative').optional(),
      fuel_liters: z.number().min(0, 'Cannot be negative').optional(),
      fuel_cost: z.number().min(0, 'Cannot be negative').optional(),
    })
    .refine((data) => data.end_odometer >= startOdometer, {
      message: `Must be ≥ start (${startOdometer.toLocaleString()} km)`,
      path: ['end_odometer'],
    })
    .refine(
      (data) => {
        const hasLiters = data.fuel_liters !== undefined && data.fuel_liters > 0;
        const hasCost = data.fuel_cost !== undefined && data.fuel_cost > 0;
        return hasLiters === hasCost;
      },
      {
        message: 'Fuel liters and fuel cost must both be provided, or both left empty',
        path: ['fuel_cost'],
      }
    );
}

export type CompleteTripInputs = z.infer<ReturnType<typeof buildCompleteTripSchema>>;

// ---------------------------------------------------------------------------
// Shared API error helper — unwraps the standard { error: { code, message,
// field } } envelope (docs/03-API-SPEC.md §2) so callers can map `field`
// onto a react-hook-form field error, or fall back to a form-level banner.
// ---------------------------------------------------------------------------

export function getApiErrorMessage(
  error: unknown,
  fallback = 'Something went wrong. Please try again.'
): { message: string; field?: string | null } {
  if (isAxiosError<ErrorEnvelope>(error) && error.response?.data?.error) {
    const { message, field } = error.response.data.error;
    return { message, field };
  }
  return { message: fallback };
}
